hail/src/main/scala/is/hail/expr/ir/PCode.scala

```diff
@@ -5,12 +5,58 @@ import is.hail.asm4s._
 import is.hail.utils._
 import is.hail.expr.types.physical._
 
-abstract class PValue {
+trait PValue {
   def pt: PType
 
   def get: PCode
 }
 
+abstract class PIndexableValue extends PValue {
+  def loadLength(): Value[Int]
+
+  def loadElement(cb: EmitCodeBuilder, i: Code[Int]): IEmitCode
+}
+
+abstract class PIndexableSettable extends PIndexableValue with PSettable {
+  def loadLength(): Value[Int]
+
+  def loadElement(cb: EmitCodeBuilder, i: Code[Int]): IEmitCode
+}
+
+object PCanonicalIndexableSettable {
+  def apply(sb: SettableBuilder, pt: PContainer, name: String): PCanonicalIndexableSettable = {
+    new PCanonicalIndexableSettable(pt,
+      sb.newSettable[Long](s"${ name }_a"),
+      sb.newSettable[Int](s"${ name }_length"),
+      sb.newSettable[Long](s"${ name }_elems_addr"))
+  }
+}
+
+class PCanonicalIndexableSettable(
+  val pt: PContainer,
+  val a: Settable[Long],
+  val length: Settable[Int],
+  val elementsAddress: Settable[Long]
+) extends PIndexableSettable {
+  def get: PIndexableCode = new PCanonicalIndexableCode(pt, a)
+
+  def loadLength(): Value[Int] = length
+
+  def loadElement(cb: EmitCodeBuilder, i: Code[Int]): IEmitCode = {
+    val iv = cb.memoize(i, "pcindval_i")
```

Comment:
A lot of the time, I expect `i` to already be a `Value`. Is there an optimization pass that removes locals that are created because we memoize a `Value`?

Reply:
Good catch.  Not yet, but it is on the TODO list.  Instead of an optimization pass (an option, although not totally trivial), my plan is to mark some variables as "const", and then we don't have to re-memoize const values.